lxd/device/gpu_physical.go

```diff
@@ -98,10 +101,209 @@ func (d *gpuPhysical) Start() (*deviceConfig.RunConfig, error) {
+	srcFDHandlers := make([]*os.File, 0)
+	defer func() {
+		for _, f := range srcFDHandlers {
+			_ = f.Close()
+		}
+	}()
+
+	for _, conf := range configDevices.UnixCharDevs {
+		if conf["source"] == "" {
+			return fmt.Errorf("The source of the unix-char device %q used for CDI is empty", conf)
+		}
+
+		_, major, minor, err := unixDeviceAttributes(conf["source"])
+		if err != nil {
+			return err
+		}
+
+		// Here putting a 'cdi.unix' prefix with 'd.name' as a device name with create an directory entry like:
+		// <lxd_var_path>/devices/<instance_name>/cdi.unix.<gpu_device_name>.<path_encoded_relative_dest_path>
+		// 'unixDeviceSetupCharNum' is already checking for dupe entries so we have no validation to do here.
+		err = unixDeviceSetupCharNum(d.state, d.inst.DevicesPath(), "cdi.unix", d.name, conf, major, minor, conf["path"], false, runConf)
+		if err != nil {
+			return err
+		}
+	}
+
+	for _, conf := range configDevices.BindMounts {
+		if conf["source"] == "" {
+			return fmt.Errorf("The source of the disk device %q used for CDI is empty", conf)
+		}
+
+		srcPath := shared.HostPath(conf["source"])
+		destPath := conf["path"]
+		relativeDestPath := strings.TrimPrefix(destPath, "/")
+
+		// This time, the created path will be like:
+		// <lxd_var_path>/devices/<instance_name>/nvidia.com.<gpu_device_name>.<path_encoded_relative_dest_path>
+		deviceName := filesystem.PathNameEncode(deviceJoinPath(string(cdiID.Vendor), d.name, relativeDestPath))
+		devPath := filepath.Join(d.inst.DevicesPath(), deviceName)
+
+		ownerShift := deviceConfig.MountOwnerShiftNone
+		if shared.IsTrue(conf["shift"]) {
+			ownerShift = deviceConfig.MountOwnerShiftDynamic
+		}
+
+		options := []string{"bind"}
+		mntOptions := shared.SplitNTrimSpace(conf["raw.mount.options"], ",", -1, true)
+		fsName := "none"
+
+		fileInfo, err := os.Stat(srcPath)
+		if err != nil {
+			return fmt.Errorf("Failed accessing source path %q: %w", srcPath, err)
+		}
+
+		fileMode := fileInfo.Mode()
+		isFile := false
+		if !fileMode.IsDir() {
+			isFile = true
+		}
+
+		f, err := os.OpenFile(srcPath, unix.O_PATH|unix.O_CLOEXEC, 0)
+		if err != nil {
+			return fmt.Errorf("Failed opening source path %q: %w", srcPath, err)
+		}
+
+		srcPath = fmt.Sprintf("/proc/self/fd/%d", f.Fd())
+		srcFDHandlers = append(srcFDHandlers, f)
+
+		// Create the devices directory if missing.
+		if !shared.PathExists(d.inst.DevicesPath()) {
+			err := os.Mkdir(d.inst.DevicesPath(), 0711)
+			if err != nil {
+				return err
+			}
+		}
+
+		// Clean any existing entry.
+		// We could have just continued because this qualifies as a duplicate case
+		// (e.g: trying to mount the same user space mount) but then if we did that
+		// we would stop the container from receiving a potentially 'updated' version
+		// of that mount (e.g: auto-update a user space lib in the container)
+		// We prefer that approach in order to keep the host and container user spaces in sync.
+		if shared.PathExists(devPath) {
```

Comment:
im not quite following this comment, please can you explain in what scenarios this occurs?

Reply:
I realize this is also a mistake from me...

I initially though the following: if we have an existing running container with a CDI device `gpu0` , what if we add an other CDI device `gpu1` while the instance is running with `lxc config device add c1 gpu1 gpu gputype=physical id=nvidia.com/gpu=1` ? This is actually permitted by LXD (I checked). So potentially that this `gpu1` will bring its own set of mounts from its CDI spec that are likely to be the same name as the `gpu0` ones but with a different content (maybe they are updated versions of these libraries because they are not passed to the container at the same point in time) .

Now, I realize that: these libs from `gpu0` are bind mounts that when updated on host are also updated in the container (though there are read only from the container). Correct? So the 'update' will happen anyway without adding a new CDI device at a different point in time... 

Basically this comment is useless because each time a CDI device is started, its bind mounts represent the common edits that are used for ALL the device so removing an existing entry (that will be created right after) is the good way to go because it ensures what is inside the container is always in sync with the host. I don't know if all of this makes sense..

Reply:
I think I follow it. Thanks.

Reply:
fixed